core/services/headtracker/orm.go

```diff
@@ -7,61 +7,65 @@ import (
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/pkg/errors"
+
 	"github.com/smartcontractkit/chainlink/core/logger"
 	"github.com/smartcontractkit/chainlink/core/services/eth"
 	"github.com/smartcontractkit/chainlink/core/services/pg"
 	"github.com/smartcontractkit/chainlink/core/utils"
 	"github.com/smartcontractkit/sqlx"
 )
 
-type ORM struct {
+type ORM interface {
+	// IdempotentInsertHead inserts a head only if the hash is new. Will do nothing if hash exists already.
+	// No advisory lock required because this is thread safe.
+	IdempotentInsertHead(ctx context.Context, head *eth.Head) error
+	// TrimOldHeads deletes heads such that only the top N block numbers remain
+	TrimOldHeads(ctx context.Context, n uint) (err error)
+	// LatestHead returns the highest seen head
+	LatestHead(ctx context.Context) (head *eth.Head, err error)
+	// LatestHeads returns the latest heads up to given limit
+	LatestHeads(ctx context.Context, limit uint) (heads []*eth.Head, err error)
+	// HeadByHash fetches the head with the given hash from the db, returns nil if none exists
+	HeadByHash(ctx context.Context, hash common.Hash) (head *eth.Head, err error)
+}
+
+type orm struct {
 	q       pg.Q
 	chainID utils.Big
 }
 
-func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.LogConfig, chainID big.Int) *ORM {
-	if db == nil {
-		panic("db may not be nil")
-	}
-	return &ORM{pg.NewQ(db, lggr, cfg), utils.Big(chainID)}
+func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.LogConfig, chainID big.Int) ORM {
+	return &orm{pg.NewQ(db, lggr, cfg), utils.Big(chainID)}
 }
 
-// IdempotentInsertHead inserts a head only if the hash is new. Will do nothing if hash exists already.
-// No advisory lock required because this is thread safe.
-func (orm *ORM) IdempotentInsertHead(ctx context.Context, h *eth.Head) error {
-	if h.EVMChainID == nil {
-		h.EVMChainID = &orm.chainID
-	} else if ((*big.Int)(h.EVMChainID)).Cmp((*big.Int)(&orm.chainID)) != 0 {
-		return errors.Errorf("head chain ID %s does not match orm chain ID %s", h.EVMChainID.String(), orm.chainID.String())
-	}
+func (orm *orm) IdempotentInsertHead(ctx context.Context, head *eth.Head) error {
+	// listener guarantees head.EVMChainID to be equal to orm.chainID
 	q := orm.q.WithOpts(pg.WithParentCtx(ctx))
 	query := `
-INSERT INTO heads (hash, number, parent_hash, created_at, timestamp, l1_block_number, evm_chain_id, base_fee_per_gas) VALUES (
-:hash, :number, :parent_hash, :created_at, :timestamp, :l1_block_number, :evm_chain_id, :base_fee_per_gas)
-ON CONFLICT (evm_chain_id, hash) DO NOTHING
-`
-	err := q.ExecQNamed(query, h)
+	INSERT INTO heads (hash, number, parent_hash, created_at, timestamp, l1_block_number, evm_chain_id, base_fee_per_gas) VALUES (
+	:hash, :number, :parent_hash, :created_at, :timestamp, :l1_block_number, :evm_chain_id, :base_fee_per_gas)
+	ON CONFLICT (evm_chain_id, hash) DO NOTHING
+	`
```

Comment:
This trailing newline seems irregular.
```suggestion
	ON CONFLICT (evm_chain_id, hash) DO NOTHING`
```